lib/vega_lite/data.ex

```diff
@@ -0,0 +1,231 @@
+      |> Vl.mark(:bar)
+
+  The above examples achieves the same results as the example below.
+
+      Vl.new()
+      |> Vl.data_from_values(data, only: ["category", "score"])
+      |> Vl.mark(:bar)
+      |> Vl.encode_field(:x, "category", type: :nominal)
+      |> Vl.encode_field(:y, "score", type: :quantitative)
+  """
+  def chart(%Vl{} = vl, data, fields) do
+    cols = columns_for(data)
+    used_fields = fields |> Keyword.values() |> used_fields()
+    root = vl |> Vl.data_from_values(data, only: used_fields)
+    for {field, col} <- fields, reduce: root do
+      acc -> encode_field(acc, cols, field, col)
+    end
+  end
+
+  def chart(data, mark, fields), do: chart(Vl.new(), data, mark, fields)
+
+  @doc """
+  Same as chart/3 but receives a valid `VegaLite` specification as a first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "With title")
+      |> Data.chart(data, :bar, x: "category", y: "score")
+
+  The above example achieves the same results as the example below.
+
+      Vl.new(title: "With title")
+      |> Vl.data_from_values(data, only: ["category", "score"])
+      |> Vl.mark(:bar)
+      |> Vl.encode_field(:x, "category", type: :nominal)
+      |> Vl.encode_field(:y, "score", type: :quantitative)
+  """
+  def chart(vl, data, mark, fields) do
+    cols = columns_for(data)
+    used_fields = fields |> Keyword.values() |> used_fields()
+    root = vl |> Vl.data_from_values(data, only: used_fields) |> enconde_mark(mark)
+
+    for {field, col} <- fields, reduce: root do
+      acc -> encode_field(acc, cols, field, col)
+    end
+  end
+
+  @doc """
+  Returns the specification of a heat map for a given data and a list of fields to be encoded.
+
+  As a specialized chart, the heatmap expects an `:x` and `:y` and optionally a `:color` and a
+  `:text` field. Defaults to `:nominal` for the axes and `:quantitative` for color and text if
+  types are not specified.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+
+  """
+  def heatmap(data, fields), do: heatmap(Vl.new(), data, fields)
+
+  @doc """
+  Same as heatmap/2, but takes a valid `VegaLite` specification as the first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "Heatmap", width: 500)
+      |> Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+  """
+  def heatmap(vl, data, fields) do
+    fields = Enum.map(fields, &heatmap_defaults/1)
+
+    if fields[:text],
+      do: annotated_heatmap(vl, data, fields),
+      else: chart(vl, data, :rect, fields)
+  end
+
+  defp annotated_heatmap(vl, data, fields) do
```

Comment:
My suggestion: I would have a function that first normalizes all `fields` to be `[field: ...] ++ _`. This way you can assume it will always have `[field: ...]` as first and you don't have to deal with list vs string everywhere below.